lib/iris/fileformats/grib/message.py

```diff
@@ -259,6 +274,16 @@ def __init__(self, message_id):
         self._message_id = message_id
         self._sections = None
 
+    def _cache_header(self):
+        """
+        Create sections and cache the header keys and values
+        as the file will be closed at some point.
+
+        """
+        for section in self.sections:
```

Comment:
As it is now, the thing called 'section' is an index.
But it's only used the context "self.sections[section]".
So I still think it's clearer to iterate over the actual section objects...
```
    for section in self.sections.values():
        for key in section.keys():
            ....
```